[rush] rush deploy does not copy over package binaries

**Is this a feature or a bug?**

- [ ] Feature
- [x] Bug

**Please describe the actual behavior.**

When executing `package.json` `scripts`, the underlying binary cannot be found because the `.bin` folder has not been copied (which I think is where `npm` looks for binaries)

example stack trace trying to call `knex`

```
/opt/deploy/apps/api $ npm run db:migrate:prod

> api@0.0.1 db:migrate:prod /opt/deploy/apps/api
> knex migrate:latest --verbose

sh: knex: not found
npm ERR! code ELIFECYCLE
npm ERR! syscall spawn
npm ERR! file sh
npm ERR! errno ENOENT
npm ERR! api@0.0.1 db:migrate:prod: `knex migrate:latest --verbose`
npm ERR! spawn ENOENT
npm ERR!
npm ERR! Failed at the api@0.0.1 db:migrate:prod script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
```

**If the issue is a bug, how can we reproduce it? Please provide detailed steps and include a GitHub branch if applicable. Your issue will get resolved faster if you can make it easy to investigate.**

1. Install any package that also installs a binary (eslint or prettier are both good choices). 
2. Create a `package.json` `script` that calls the binary. For example, create a script called `prettier` with the following content: `prettier src/**/* --write`
3. Confirm that `npm run prettier` and/or `rushx prettier` works while inside the app directly.
4. `rush build` && `rush deploy`
5. `cd common/deploy/apps/api` and run `npm run prettier` or `rushx prettier` 

From gitter, [this is how `rush install` does it](https://github.com/microsoft/rushstack/blob/040dc7ae65fb0102345570a967e5d490b4c166c3/apps/rush-lib/src/logic/pnpm/PnpmLinkManager.ts#L284)

**What is the expected behavior?**

Packages that install binaries should be included / symlinked / copied during the deploy process.

**If this is a bug, please provide the tool version, Node.js version, and OS.**

* **Tool:** pnpm
* **Tool Version:** 5.2.6
* **Node Version:** 12.16.3
  * **Is this a LTS version?** yes
  * **Have you tested on a LTS version?** yes
* **OS:** MacOS and Alpine linux

--- a/apps/rush-lib/package.json
+++ b/apps/rush-lib/package.json
@@ -18,7 +18,7 @@
   },
   "license": "MIT",
   "dependencies": {
-    "@pnpm/link-bins": "~5.1.0",
+    "@pnpm/link-bins": "~5.3.7",
     "@rushstack/node-core-library": "workspace:*",
     "@rushstack/package-deps-hash": "workspace:*",
     "@rushstack/stream-collator": "workspace:*",

--- a/apps/rush-lib/src/cli/actions/DeployAction.ts
+++ b/apps/rush-lib/src/cli/actions/DeployAction.ts
@@ -68,7 +68,7 @@ export class DeployAction extends BaseRushAction {
 
   protected async run(): Promise<void> {
     const deployManager: DeployManager = new DeployManager(this.rushConfiguration);
-    deployManager.deploy(
+    await deployManager.deploy(
       this._project.value,
       this._scenario.value,
       !!this._overwrite.value,

--- a/apps/rush-lib/src/cli/actions/DeployAction.ts
+++ b/apps/rush-lib/src/cli/actions/DeployAction.ts
@@ -68,7 +68,7 @@ export class DeployAction extends BaseRushAction {
 
   protected async run(): Promise<void> {
     const deployManager: DeployManager = new DeployManager(this.rushConfiguration);
-    await deployManager.deploy(
+    await deployManager.deployAsync(
       this._project.value,
       this._scenario.value,
       !!this._overwrite.value,

--- a/apps/rush-lib/src/logic/deploy/DeployManager.ts
+++ b/apps/rush-lib/src/logic/deploy/DeployManager.ts
@@ -5,6 +5,7 @@ import * as colors from 'colors';
 import * as path from 'path';
 import * as resolve from 'resolve';
 import * as npmPacklist from 'npm-packlist';
+import pnpmLinkBins from '@pnpm/link-bins';
 
 // (Used only by the legacy code fragment in the resolve.sync() hook below)
 import * as fsForResolve from 'fs';
@@ -44,7 +45,18 @@ declare module 'npm-packlist' {
 export interface IDeployMetadataJson {
   scenarioName: string;
   mainProjectName: string;
+  projects: IProject[];
   links: ILinkInfo[];
+}
+
+/**
+ * Represents a Rush project to be deployed.
+ */
+interface IProject {
+  /**
+   * This path is relative to the deploy folder.
+   */
+  path: string;
 }
 
 /**
@@ -461,8 +473,19 @@ export class DeployManager {
     const deployMetadataJson: IDeployMetadataJson = {
       scenarioName: path.basename(deployState.scenarioFilePath),
       mainProjectName: deployState.mainProjectName,
+      projects: [],
       links: []
     };
+
+    deployState.folderInfosByPath.forEach((folderInfo) => {
+      if (!folderInfo.isRushProject) {
+        return;
+      }
+
+      deployMetadataJson.projects.push({
+        path: this._remapPathForDeployMetadata(folderInfo.folderPath, deployState)
+      });
+    });
 
     // Remap the links to be relative to target folder
     for (const absoluteLinkInfo of deployState.symlinkAnalyzer.reportSymlinks()) {
@@ -479,7 +502,28 @@ export class DeployManager {
     });
   }
 
-  private _prepareDeployment(deployState: IDeployState): void {
+  private async _makeBinLinks(deployState: IDeployState): Promise<void> {
+    for (const [, folderInfo] of deployState.folderInfosByPath) {
+      if (!folderInfo.isRushProject) {
+        return;
+      }
+
+      const deployedPath: string = this._remapPathForDeployMetadata(folderInfo.folderPath, deployState);
+      const projectFolder: string = path.join(deployState.targetRootFolder, deployedPath, 'node_modules');
+      const projectBinFolder: string = path.join(
+        deployState.targetRootFolder,
+        deployedPath,
+        'node_modules',
+        '.bin'
+      );
+
+      await pnpmLinkBins(projectFolder, projectBinFolder, {
+        warn: (msg: string) => console.warn(colors.yellow(msg))
+      });
+    }
+  }
+
+  private async _prepareDeployment(deployState: IDeployState): Promise<void> {
     // Calculate the set with additionalProjectsToInclude
     const includedProjectNamesSet: Set<string> = new Set();
     this._collectAdditionalProjectsToInclude(
@@ -540,18 +584,20 @@ export class DeployManager {
           throw new InternalError('Target does not exist: ' + JSON.stringify(linkToCopy, undefined, 2));
         }
       }
+
+      await this._makeBinLinks(deployState);
     }
   }
 
   /**
    * The main entry point for performing a deployment.
    */
-  public deploy(
+  public async deploy(
     mainProjectName: string | undefined,
     scenarioName: string | undefined,
     overwriteExisting: boolean,
     targetFolderParameter: string | undefined
-  ): void {
+  ): Promise<void> {
     const scenarioFilePath: string = DeployScenarioConfiguration.getConfigFilePath(
       scenarioName,
       this._rushConfiguration
@@ -625,7 +671,7 @@ export class DeployManager {
 
     console.log();
 
-    this._prepareDeployment(deployState);
+    await this._prepareDeployment(deployState);
 
     console.log('\n' + colors.green('The operation completed successfully.'));
   }

--- a/apps/rush-lib/src/logic/deploy/DeployManager.ts
+++ b/apps/rush-lib/src/logic/deploy/DeployManager.ts
@@ -502,7 +502,7 @@ export class DeployManager {
     });
   }
 
-  private async _makeBinLinks(deployState: IDeployState): Promise<void> {
+  private async _makeBinLinksAsync(deployState: IDeployState): Promise<void> {
     for (const [, folderInfo] of deployState.folderInfosByPath) {
       if (!folderInfo.isRushProject) {
         return;
@@ -523,7 +523,7 @@ export class DeployManager {
     }
   }
 
-  private async _prepareDeployment(deployState: IDeployState): Promise<void> {
+  private async _prepareDeploymentAsync(deployState: IDeployState): Promise<void> {
     // Calculate the set with additionalProjectsToInclude
     const includedProjectNamesSet: Set<string> = new Set();
     this._collectAdditionalProjectsToInclude(
@@ -585,14 +585,14 @@ export class DeployManager {
         }
       }
 
-      await this._makeBinLinks(deployState);
+      await this._makeBinLinksAsync(deployState);
     }
   }
 
   /**
    * The main entry point for performing a deployment.
    */
-  public async deploy(
+  public async deployAsync(
     mainProjectName: string | undefined,
     scenarioName: string | undefined,
     overwriteExisting: boolean,
@@ -671,7 +671,7 @@ export class DeployManager {
 
     console.log();
 
-    await this._prepareDeployment(deployState);
+    await this._prepareDeploymentAsync(deployState);
 
     console.log('\n' + colors.green('The operation completed successfully.'));
   }

--- a/apps/rush-lib/src/logic/deploy/DeployManager.ts
+++ b/apps/rush-lib/src/logic/deploy/DeployManager.ts
@@ -479,6 +479,12 @@ export class DeployManager {
 
     deployState.folderInfosByPath.forEach((folderInfo) => {
       if (!folderInfo.isRushProject) {
+        // It's not a Rush project
+        return;
+      }
+
+      if (!deployState.foldersToCopy.has(folderInfo.folderPath)) {
+        // It's not something we crawled
         return;
       }
 

--- a/apps/rush-lib/src/logic/deploy/DeployManager.ts
+++ b/apps/rush-lib/src/logic/deploy/DeployManager.ts
@@ -45,14 +45,14 @@ declare module 'npm-packlist' {
 export interface IDeployMetadataJson {
   scenarioName: string;
   mainProjectName: string;
-  projects: IProject[];
+  projects: IProjectInfoJson[];
   links: ILinkInfo[];
 }
 
 /**
- * Represents a Rush project to be deployed.
+ * Part of the deploy-matadata.json file format. Represents a Rush project to be deployed.
  */
-interface IProject {
+interface IProjectInfoJson {
   /**
    * This path is relative to the deploy folder.
    */

--- a/common/changes/@microsoft/rush/fix-1982_2020-07-08-17-13.json
+++ b/common/changes/@microsoft/rush/fix-1982_2020-07-08-17-13.json
@@ -2,7 +2,7 @@
   "changes": [
     {
       "packageName": "@microsoft/rush",
-      "comment": "[rush-lib] Copy over package binaries during deploy (#1982)",
+      "comment": "Fix an issue where package binaries where not created by \"rush deploy\" (#1982)",
       "type": "none"
     }
   ],

--- a/common/changes/@microsoft/rush/fix-1982_2020-07-08-17-13.json
+++ b/common/changes/@microsoft/rush/fix-1982_2020-07-08-17-13.json
@@ -1,0 +1,11 @@
+{
+  "changes": [
+    {
+      "packageName": "@microsoft/rush",
+      "comment": "[rush-lib] Copy over package binaries during deploy (#1982)",
+      "type": "none"
+    }
+  ],
+  "packageName": "@microsoft/rush",
+  "email": "willsoto@users.noreply.github.com"
+}

--- a/common/config/rush/repo-state.json
+++ b/common/config/rush/repo-state.json
@@ -1,5 +1,5 @@
 // DO NOT MODIFY THIS FILE. It is generated and used by Rush.
 {
-  "pnpmShrinkwrapHash": "2ac6991c2d70af3226ed6ee10edd64632a686ce9",
+  "pnpmShrinkwrapHash": "4adc15ab5bc85ff3c787140b2ba9d90ecd2f9867",
   "preferredVersionsHash": "334ea62b6a2798dcf80917b79555983377e7435e"
 }

